collectd/collectd.go

```diff
@@ -0,0 +1,220 @@
+}
+
+type Server struct {
+	mu   sync.Mutex
+	wg   sync.WaitGroup
+	done chan struct{} // close notification
+	conn *net.UDPConn
+
+	writer      SeriesWriter
+	Database    string
+	typesdb     gollectd.Types
+	typesdbpath string
+}
+
+func NewServer(w SeriesWriter, typesDBPath string) *Server {
+	s := &Server{}
+
+	s.writer = w
+	s.typesdbpath = typesDBPath
+	s.typesdb = make(gollectd.Types)
+
+	return s
+}
+
+func (s *Server) ListenAndServe(iface string) error {
+	if iface == "" { // Make sure we have an address
+		return ErrBindAddressRequired
+	} else if s.Database == "" { // Make sure they have a database
+		return ErrDatabaseNotSpecified
+	}
+
+	var err error
+	s.typesdb, err = gollectd.TypesDBFile(s.typesdbpath)
+	if err != nil {
+		return ErrCouldNotParseTypesDBFile
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", iface)
+	if err != nil {
+		return ErrResolveUDPAddr
+	}
+
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		return ErrListenUDP
+	}
+	s.conn = conn
+
+	// Create a new close notification channel.
+	done := make(chan struct{}, 0)
+	s.done = done
+
+	s.wg.Add(1)
+	go s.serve(conn)
+
+	return nil
+}
+
+func (s *Server) serve(conn *net.UDPConn) {
+	defer s.wg.Done()
+
+	// From https://collectd.org/wiki/index.php/Binary_protocol
+	//   1024 bytes (payload only, not including UDP / IP headers)
+	//   In versions 4.0 through 4.7, the receive buffer has a fixed size
+	//   of 1024 bytes. When longer packets are received, the trailing data
+	//   is simply ignored. Since version 4.8, the buffer size can be
+	//   configured. Version 5.0 will increase the default buffer size to
+	//   1452 bytes (the maximum payload size when using UDP/IPv6 over
+	//   Ethernet).
+	buffer := make([]byte, 1452)
+
+	for {
+		select {
+		case <-s.done:
+			s.conn.Close()
+			return
+		default:
+			n, _, err := conn.ReadFromUDP(buffer)
+			if err != nil || n == 0 {
+				log.Printf("Collectd ReadFromUDP error: %s", err)
+				return
+			}
+
+			// Read in data in a separate goroutine.
+			s.wg.Add(1)
+			go s.handleMessage(buffer[:n])
+		}
+	}
+}
+
+func (s *Server) handleMessage(buffer []byte) {
+	defer s.wg.Done()
+
+	packets, err := gollectd.Packets(buffer, s.typesdb)
+	if err != nil {
+		log.Printf("Collectd parse error: %s", err)
+		return
+	}
+
+	for _, packet := range *packets {
+		metrics := Unmarshal(&packet)
+		for _, m := range metrics {
+			// Convert metric to a field value.
+			var values = make(map[string]interface{})
+			values[m.Name] = m.Value
+
+			err := s.writer.WriteSeries(s.Database, "", m.Name, m.Tags, m.Timestamp, values)
+			if err != nil {
+				log.Printf("Collectd cannot write data: %s", err)
+				continue
+			}
+		}
+	}
+}
+
+// Close shuts down the server's listeners.
+func (s *Server) Close() error {
+	// Notify other goroutines of shutdown.
+	s.mu.Lock()
+	if s.done == nil {
+		s.mu.Unlock()
+		return ErrServerClosed
+	}
+
+	s.mu.Unlock()
+	close(s.done)
```

Comment:
Can you close `s.done` and set it to `nil` within the context of the lock? Otherwise you can technically have a race condition if two goroutines call `Server.Close()` at the same time.